### PR TITLE
Dynamic Form Manager

### LINKED ISSUE #26: Dyn Form Manager
# Field Types

| Name | DB Type | Validations |
| --- | --- | --- |
| Salutation | int: 1 |  |
| Title | string: 255 |  |
| Firstname | string: 255 |  |
| MiddleName | string: 255 |  |
| LastName | string: 255 |  |
| Birthday | dateTime | dateTime |
| Email | string: 255 | Email |
| Street | string: 255 |  |
| Number | string: 255 |  |
| ZIP | string: 255 |  |
| Region / State | string: 255 |  |
| Country | string: 2 | Country |
| Newsletter | bool: 1 |  |
| Company | string: 255 |  |
| Function | string: 255 |  |
| Typ of Company | string: 255 |  |
| Phone | string: 255 |  |
| Fax | string: 255 |  |
| Notice | text |  |
| From-To | dateTime/dateTime | dateTime/dateTime |
| Select | string:255 |  |
| Multiple Select | text |  |
| Checkbox | bool: 1 |  |
| Checkboxes | text |  |
| Radio | string:255 |  |
| Simple Textfield | string: 255 |  |
| Textarea | text |  |
| Date | date | date |
| Time | time | time |
| File Upload | int -> media |  |
# Special Field Types

| Name | Description | Syntax |
| --- | --- | --- |
| Captcha | only when captcha bundle is available |  |
| Headline | Text Headline | `<div class="form-headline">Headline</div>` |
| Text | Text Editor | `<div class="form-text"><p>Test</p></div>` |
| Spacer | Just a space | `<div class="form-spacer"></div>` |
# Available Validations

| Name | Field Types |
| --- | --- |
| NotBlank ( by checkbox) | all |
| MinLength | textfields |
| MaxLength | textfields |
# Template Fields

| Name | Content Type |
| --- | --- |
| Subject | text_line |
| From Name | text_line |
| From Email | text_line |
| To Name | text_line |
| Thank you Text | text_editor |
| Thank you Email | text_editor |
# Save Fields

| Name | DB Type |
| --- | --- |
| uuid | string: 255 |
| webspaceKey | string: 255 |
| locale | string: 5 |
# Features
- List View
- CSV Export

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | fixes #26 |
| Related issues/PRs | #39, #41 |
| License | MIT |

**TODO List:**
- [ ] Finish Content Type
  - [ ] Add the selection of the form in the overlay
  - [x] Add simple autocomplete
  - [x] Create Symfony Form
  - [x] Create dist files
  - [x] Localize Content Type API (?locale=de)
- [x] Form Edit
  - [x] Remove Blocks
  - [x] Sort Blocks
  - [x] Load Form Queries
  - [x] Choices
- [ ] Save Form
  - [x] Form Handling in Content Type
  - [x] Email Handling
  - [x] Save Data
  - [x] Save Media to Dynamic enitity
  - [x] Save Form id to Dynamic enitity
  - [ ] Export List
  - [x] File Upload
  - [x] System Collection

### CODE REVIEW

File: Content/Types/FormSelect.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                }
+            }
+
+            // Create Form Type
+            $formType = new DynamicFormType(
+                $formEntity,
+                $locale,
+                $property->getName(),
+                $property->getStructure()->getView(),
+                $this->systemCollectionManager->getSystemCollection()
```
Comment (alexander-schranz) [inline Content/Types/FormSelect.php]: @thomasduenser is not the corect key missing here?

Comment (thomasduenser) [inline Content/Types/FormSelect.php]: @alexander-schranz Yes, I've already fixed that in the next MR!

File: Entity/Dynamic.php [hunk 2]
```diff
@@ -2,118 +2,288 @@
...
+     * Returns an array of all constants.
      *
-     * @return string
+     * @return array
      */
-    public function getForm()
+    public static function getConstants()
     {
-        return $this->form;
+        $reflectionClass = new ReflectionClass(__CLASS__);
```
Comment (alexander-schranz) [inline Entity/Dynamic.php]: @thomasduenser avoid reflection as it is a performance bottleneck. create a static variable.

``` php
public static $TYPES = [
    self::TYPE_SPACER,
    self::TYPE_FREE_TEXT,
    self::TYPE_SALUTATION,
    self::TYPE_TITLE,
    self::TYPE_FIRST_NAME,
    self::TYPE_LAST_NAME,
    self::TYPE_EMAIL,
    self::TYPE_PHONE,
    self::TYPE_FAX,
    self::TYPE_STREET,
    self::TYPE_ZIP,
    self::TYPE_CITY,
    self::TYPE_STATE,
    self::TYPE_COUNTRY,
    self::TYPE_FUNCTION,
    self::TYPE_COMPANY,
    self::TYPE_TEXT,
    self::TYPE_TEXTAREA,
    self::TYPE_DATE,
    self::TYPE_HEADLINE,
    self::TYPE_ATTACHMENT,
    self::TYPE_CHECKBOX,
    self::TYPE_CHECKBOX_MULTIPLE,
    self::TYPE_DROPDOWN,
    self::TYPE_DROPDOWN_MULTIPLE,
    self::TYPE_RADIO_BUTTONS,
];
```

File: Entity/Dynamic.php [hunk 3]
```diff
@@ -2,118 +2,288 @@
...
+    const TYPE_FUNCTION = 'function';
+    const TYPE_COMPANY = 'company';
+    const TYPE_TEXT = 'text';
+    const TYPE_TEXTAREA = 'textarea';
+    const TYPE_DATE = 'date';
+    const TYPE_HEADLINE = 'headline';
+    const TYPE_ATTACHMENT = 'attachment';
+    const TYPE_CHECKBOX = 'checkbox';
+    const TYPE_CHECKBOX_MULTIPLE = 'checkboxMultiple';
+    const TYPE_DROPDOWN = 'dropdown';
```
Comment (alexander-schranz) [inline Entity/Dynamic.php:32]: @thomasduenser why not call it `select`?

Comment (thomasduenser) [inline Entity/Dynamic.php:32]: @alexander-schranz Because I had problems, as `select` is a reserved SQL word.

File: Entity/Dynamic.php (hunk 2 above)
Comment (thomasduenser) [inline Entity/Dynamic.php]: @alexander-schranz Thanks for the info, will change it!

**Pull request merged**